### ISSUE TITLE
Make algorithm specific EVP_PKEY_CTX setters and getters more available

### ISSUE THREAD

**Issue by levitte** [openssl/openssl#13550]
Algorithm specific EVP_PKEY_CTX setters and getters have generally only been made available when the algorithm in question is enabled.  So for example, when `no-rsa` is configured, `EVP_PKEY_CTX_set_rsa_padding()` and `EVP_PKEY_CTX_get_rsa_padding()` aren't available.

As long as only methods that are built-in in libcrypto are available, that's fine, because it reflects what implementations are available.
However, with ENGINEs and providers that may implement algorithms regardless of OpenSSL configuration, having those functions disabled becomes more questionable.

This has been changed already for DSA, because of #13529, fixed by #13530.
The rest of the key types for which we have this sort of functionality needs to be updated the same way: DH, RSA and EC_KEY

**Comment by levitte** [comment]
I've started this work on RSA.  That one turned out to be fairly complex.
DSA was on the other hand a fairly easy job.

**Comment by mattcaswell** [comment]
Note that in spite of the many `OPENSSL_NO_RSA` guards that we have, `no-rsa` isn't actually a valid option.

**Comment by levitte** [comment]
> Note that in spite of the many `OPENSSL_NO_RSA` guards that we have, `no-rsa` isn't actually a valid option.

Oh, thanks for the reminder!  The easy path would then be to simply drop those guards everywhere...  considering that we don't have them in half the spots where they would otherwise be, anyone trying to configure with `-DOPENSSL_NO_RSA` will end up with a deeply inconsistent library.

**Comment by t8m** [comment]
Yep. No need to expand the scope of 3.0 by making no-rsa work.

**Comment by levitte** [comment]
I grew impatient, so err, #13589 does DH and EC.

**Comment by h-vetinari** [comment]
https://github.com/openssl/openssl/commit/c829c23b67308ad8e8ab677c78db1d5151106c3c says `Partially fixes openssl#13550`. What's missing resp. should this stay open?

**Comment by levitte** [comment]
> What's missing resp. should this stay open?

That commit fixed this issue specifically for DH.  You also have 565b339, which fixes it for EC_KEY.  You also have 888bdbfd398c967daaa00cf6b3d104f0e3d26865 (from #13530), which did the same for, but that was before we raised this issue.

We also concluded that there's no real need to do this work for RSA, at least in this form.  We do need to clean away the OPENSSL_NO_RSA guards, though, as they aren't supported, but that's a different issue.

So as far as I can see, this issue has been dealt with as completely as can be asked for and should remain closed.
